hadoop-ozone/recon/src/main/java/org/apache/hadoop/ozone/recon/api/NSSummaryEndpoint.java

```diff
@@ -0,0 +1,1061 @@
+      List<OmBucketInfo> buckets = listBucketsUnderVolume(names[0]);
+      String volKey = omMetadataManager.getVolumeKey(names[0]);
+      OmVolumeArgs volumeArgs =
+          omMetadataManager.getVolumeTable().getSkipCache(volKey);
+      long quotaInBytes = volumeArgs.getQuotaInBytes();
+      long quotaUsedInBytes = 0L;
+
+      // Get the total data size used by all buckets
+      for (OmBucketInfo bucketInfo: buckets) {
+        long bucketObjectId = bucketInfo.getObjectID();
+        quotaUsedInBytes += getTotalSize(bucketObjectId);
+      }
+      quotaUsageResponse.setQuota(quotaInBytes);
+      quotaUsageResponse.setQuotaUsed(quotaUsedInBytes);
+    } else if (type == EntityType.BUCKET) {
+      String bucketKey = omMetadataManager.getBucketKey(names[0], names[1]);
+      OmBucketInfo bucketInfo = omMetadataManager
+          .getBucketTable().getSkipCache(bucketKey);
+      long bucketObjectId = bucketInfo.getObjectID();
+      long quotaInBytes = bucketInfo.getQuotaInBytes();
+      long quotaUsedInBytes = getTotalSize(bucketObjectId);
+      quotaUsageResponse.setQuota(quotaInBytes);
+      quotaUsageResponse.setQuotaUsed(quotaUsedInBytes);
+    } else if (type == EntityType.UNKNOWN) {
+      quotaUsageResponse.setResponseCode(ResponseStatus.PATH_NOT_FOUND);
+    } else { // directory and key are not applicable for this request
+      quotaUsageResponse.setResponseCode(
+          ResponseStatus.TYPE_NOT_APPLICABLE);
+    }
+    return Response.ok(quotaUsageResponse).build();
+  }
+
+  /**
+   * Endpoint that returns aggregate file size distribution under a path.
+   * @param path request path
+   * @return File size distribution response
+   * @throws IOException
+   */
+  @GET
+  @Path("/dist")
+  public Response getFileSizeDistribution(@QueryParam("path") String path)
+      throws IOException {
+
+    if (path == null || path.length() == 0) {
+      return Response.status(Response.Status.BAD_REQUEST).build();
+    }
+
+    FileSizeDistributionResponse distResponse =
+        new FileSizeDistributionResponse();
+    if (!isInitializationComplete()) {
+      distResponse.setStatus(ResponseStatus.INITIALIZING);
+      return Response.ok(distResponse).build();
+    }
+
+    String normalizedPath = normalizePath(path);
+    String[] names = parseRequestPath(normalizedPath);
+    EntityType type = getEntityType(normalizedPath, names);
+
+    switch (type) {
+    case ROOT:
+      List<OmBucketInfo> allBuckets = listBucketsUnderVolume(null);
+      int[] fileSizeDist = new int[ReconConstants.NUM_OF_BINS];
+
+      // accumulate file size distribution arrays from all buckets
+      for (OmBucketInfo bucket : allBuckets) {
+        long bucketObjectId = bucket.getObjectID();
+        int[] bucketFileSizeDist = getTotalFileSizeDist(bucketObjectId);
+        // add on each bin
+        for (int i = 0; i < ReconConstants.NUM_OF_BINS; ++i) {
+          fileSizeDist[i] += bucketFileSizeDist[i];
+        }
+      }
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

Similar code fragments were detected in the same file at the following lines: 532:539, 547:554.  
Refactoring can help improve code maintainability. Consider reducing duplicate code by extracting it into a separate method. You can then replace duplicated code with calls to this new method.